src/Formatter/JunitFormatter.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\BackwardCompatibility\Formatter;
+
+use Psl\Str;
+use Roave\BackwardCompatibility\Changes;
+use Roave\BackwardCompatibility\Git\CheckedOutRepository;
+use Symfony\Component\Console\Output\OutputInterface;
+
+use function count;
+use function htmlspecialchars;
+use function implode;
+use function sprintf;
+use function trim;
+
+use const ENT_COMPAT;
+use const ENT_XML1;
+
+class JunitFormatter implements OutputFormatter
+{
+    public function __construct(
+        private OutputInterface $output,
+        private CheckedOutRepository $basePath,
+    ) {
+    }
+
+    public function write(Changes $changes): void
+    {
+        $basePath = $this->basePath->__toString() . '/';
+
+        $changeCount = count($changes);
+
+        $testcases = [];
+
+        foreach ($changes as $change) {
+            $filename = $change->file === null ? null : Str\replace($change->file, $basePath, '');
+
+            $name = implode(':', [
+                $this->escape($filename ?? ''),
+                $this->escape((string) ($change->line ?? '')),
+                $this->escape((string) ($change->column ?? '')),
+            ]);
+
+            $testcases[] = sprintf(
```

Comment:
Note: the output should probably be streamed.

A crash mid-way would make this reporter really annoying to work with :thinking:

Reply:
would streaming an xml document be useful? would interacting with a partial xml document ever be done? most xml parsers would fail i imagine

it wouldn't be too difficult to output line by line however, but if i work with `DOMDocument` instead the ability to stream it would be removed... i think?

Reply:
Streaming useful when this crashes mid-way.

But then... I noticed that we need to `count()` the results anyway, which defeats the streaming, I guess? :thinking: 

The problem here is that analysis can take between seconds and minutes, and no output for multiple minutes can be problematic for some CI environments.

Reply:
i guess that's why the github-actions output is the way that it is; it allows for partial and streaming output in a way that json / xml does not allow, as each line stands on its own

judging by the OutputFormatter interface, all the results are gathered into `Changes` before being passed to the formatter it looks like? or am i wrong there and multiple calls to `write()` can be made in one analysis?

Reply:
i've been thinking about this and the common use case of this format would be redirecting it to a file right? 

`php vendor/bin/backwards-compatibility-check --format=junit > junit.xml`
that means the CI would not see any output from the process anyway due to the redirection.

this is kind of what i meant by a partial xml document, if the output was streamed line by line to a file and the process crashed, the produced file would be incomplete and lacking the closing `</testsuite>` tag, and be invalid xml.


> judging by the OutputFormatter interface, all the results are gathered into Changes before being passed to the formatter it looks like? or am i wrong there and multiple calls to write() can be made in one analysis? 

upon looking into this it appears that all `Changes` are provided to the formatter in one go and it is not called multiple times, so i'm not sure there is a lot of value in streaming the output of the junit formatter?

however i might be missing something, so please let me know if i am.